### PR TITLE
Fieldflow plugin - Initial commit, transform a basic list into a selectable list.

### PR DESCRIPTION
**Pull request by duboisp**
Note, the following working example link are based on build 4154900

## English working example:
* [Field flow](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/demos/fieldflow/fieldflow-en.html)
* [Field flow basic configuration](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/demos/fieldflow/basic-en.html)
* [Field flow advanced](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/demos/fieldflow/advanced-en.html)

## French working example:
* [Déroulement de champs](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/demos/fieldflow/fieldflow-fr.html)
* [Déroulement de champs, configuration](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/demos/fieldflow/basic-fr.html)
* [Déroulement de champs, avancé](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/demos/fieldflow/advanced-fr.html)

## Documentation:
* [Field flow Documentation](http://universallabs.org/wet/labs/wb-fieldflow/dist-0.3/docs/ref/fieldflow/fieldflow-en.html)

## i18n temporary workaround
Currently the plugin only support English and French and use a work-around for i18n.

The following text will need to be translated in all language supported by WET, then added to the cvs, then I will be able to update the plugin to fully support the i18n of WET.

* **Button**
  * Continue
  * Allez
* **Default selected item**
  * Make your selection...
  * Sélectionnez dans la liste...
* **Required**
  * required
  * obligatoire

----
cc/ @TouficS

### CODE REVIEW

File: src/plugins/fieldflow/fieldflow.js [hunk 1]
```diff
@@ -0,0 +1,1406 @@
...
+	// Do the action upon form submission
+	pushData( $( data.provEvt ), submitJQData, data, true );
+} );
+
+/*
+ * Redirection action
+ *
+ * Add a parameter to the detination URL like "?name=value"
+ */
+$document.on( "redir." + submitEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 2]
```diff
@@ -0,0 +1,1406 @@
...
+		form.setAttribute( "action", url );
+	}
+} );
+
+/*
+ * Querying action
+ *
+ * Add a parameter to the detination URL like "?name=value"
+ */
+$document.on( "query." + actionEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 3]
```diff
@@ -0,0 +1,1406 @@
...
+	$selectElm.attr( "data-" + flagOptValueData, flagOptValueData );
+} );
+
+/* Perfoming AJAX action
+ *
+ * @elm : Element that triggered this action
+ * @dt: Data structure defining this ajax loading configuration
+ *
+ */
+$document.on( "ajax." + actionEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 4]
```diff
@@ -0,0 +1,1406 @@
...
+		$( event.target ).trigger( "ajax." + submitEvent, data );
+	}
+} );
+
+/*
+ * Ajax action
+ *
+ * Add a parameter to the detination URL like "?name=value"
+ */
+$document.on( "ajax." + submitEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 5]
```diff
@@ -0,0 +1,1406 @@
...
+					.filter( ":not(#" + updtElm.id + " .wb-init .wb-init)" )
+						.trigger( "timerpoke.wb" );
+			updtElm.removeAttribute( "data-trigger-wet" );
+		}
+	} );
+	$container.trigger( "wb-update.wb-data-ajax" );
+} );
+
+// Toggle an items
+$document.on( "toggle." + actionEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 6]
```diff
@@ -0,0 +1,1406 @@
...
+		$( event.target ).trigger( "toggle." + submitEvent, data );
+	}
+} );
+
+/*
+ * Toggle action
+ *
+ * Add a parameter to the detination URL like "?name=value"
+ */
+$document.on( "toggle." + submitEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 7]
```diff
@@ -0,0 +1,1406 @@
...
+	if ( !data.live ) {
+		data.preventSubmit = true;
+		pushData( $( data.provEvt ), submitJQData, data );
+	} else {
+		$( event.target ).trigger( "addClass." + submitEvent, data );
+	}
+} );
+
+// addClass action
+$document.on( "addClass." + submitEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 8]
```diff
@@ -0,0 +1,1406 @@
...
+	if ( !data.source || !data.class ) {
+		return;
+	}
+	$( data.source ).addClass( data.class );
+} );
+
+// removeClass action
+$document.on( "removeClass." + actionEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 9]
```diff
@@ -0,0 +1,1406 @@
...
+	if ( !data.live ) {
+		data.preventSubmit = true;
+		pushData( $( data.provEvt ), submitJQData, data );
+	} else {
+		$( event.target ).trigger( "removeClass." + submitEvent, data );
+	}
+} );
+
+// removeClass action
+$document.on( "removeClass." + submitEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

File: src/plugins/fieldflow/fieldflow.js [hunk 10]
```diff
@@ -0,0 +1,1406 @@
...
+	if ( !data.source || !data.class ) {
+		return;
+	}
+	$( data.source ).removeClass( data.class );
+} );
+
+// Insert a control next to it
+$document.on( "append." + actionEvent, selector, function( event, data ) {
```
Comment (LaurentGoderre) [inline src/plugins/fieldflow/fieldflow.js]: Collapse function with previous one and use `switch` to reduce code duplication

**Review by LaurentGoderre — CHANGES_REQUESTED** [review]

**Comment by duboisp** [comment]
@LaurentGoderre as you asked, I collapsed all the function with previous one and now it uses a ```switch```. Thanks :+1:

**Comment by LaurentGoderre** [comment]
Much cleaner and leaner!

**Comment by duboisp** [comment]
To be moved into the gc web theme, as this plugin execute business logic.

**Comment by duboisp** [comment]
This plugin is merged in GC Web
See working example and docs here: http://wet-boew.github.io/themes-dist/GCWeb/index-en.html

**Pull request closed without merging**